lighthouse-core/gather/gatherers/full-page-screenshot.js

```diff
@@ -0,0 +1,132 @@
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const Gatherer = require('./gatherer.js');
+
+// JPEG quality setting
+// Exploration and examples of reports using different quality settings: https://docs.google.com/document/d/1ZSffucIca9XDW2eEwfoevrk-OTl7WQFeMf0CgeJAA8M/edit#
+const FULL_PAGE_SCREENSHOT_QUALITY = 30;
+// Maximum screenshot height in Chrome https://bugs.chromium.org/p/chromium/issues/detail?id=770769
+const MAX_SCREENSHOT_HEIGHT = 16384;
+// Maximum data URL size in Chrome https://bugs.chromium.org/p/chromium/issues/detail?id=69227
+const MAX_DATA_URL_SIZE = 2 * 1024 * 1024 - 1;
+
+/**
+ * @param {string} str
+ */
+function snakeCaseToCamelCase(str) {
+  return str.replace(/(-\w)/g, m => m[1].toUpperCase());
+}
+
+class FullPageScreenshot extends Gatherer {
+  /**
+   * @param {LH.Gatherer.PassContext} passContext
+   * @param {number} maxScreenshotHeight
+   * @return {Promise<LH.Artifacts.FullPageScreenshot>}
+   */
+  async _takeScreenshot(passContext, maxScreenshotHeight) {
+    const driver = passContext.driver;
+    const metrics = await driver.sendCommand('Page.getLayoutMetrics');
+    const width = await driver.evaluateAsync('window.innerWidth');
+    const deviceScaleFactor = await driver.evaluateAsync('window.devicePixelRatio');
+    const height = Math.min(metrics.contentSize.height, maxScreenshotHeight);
+
+    await driver.sendCommand('Emulation.setDeviceMetricsOverride', {
+      mobile: passContext.baseArtifacts.TestedAsMobileDevice,
+      height,
+      screenHeight: height,
+      width,
+      screenWidth: width,
+      deviceScaleFactor,
+    });
+
+    const result = await driver.sendCommand('Page.captureScreenshot', {
+      format: 'jpeg',
+      quality: FULL_PAGE_SCREENSHOT_QUALITY,
+    });
+    const data = 'data:image/jpeg;base64,' + result.data;
+
+    return {
+      data,
+      width,
+      height,
+    };
+  }
+
+  /**
+   * @param {LH.Gatherer.PassContext} passContext
+   * @return {Promise<LH.Artifacts.FullPageScreenshot | null>}
+   */
+  async afterPass_(passContext) {
+    let screenshot = await this._takeScreenshot(passContext, MAX_SCREENSHOT_HEIGHT);
+
+    if (screenshot.data.length > MAX_DATA_URL_SIZE) {
+      // Hitting the data URL size limit is rare, it only happens for pages on tall
+      // desktop sites with lots of images.
+      // So just cutting down the height a bit usually fixes the issue.
+      screenshot = await this._takeScreenshot(passContext, 5000);
+      if (screenshot.data.length > MAX_DATA_URL_SIZE) {
+        passContext.LighthouseRunWarnings.push('Full page screenshot is too big.');
+        return null;
```

Comment:
any discussion earlier if this should be an error case instead of `null`? I can see the case for `null` and audit `notApplicable`, but it also seems like we usually expect this to complete, so it's a gatherer error (and therefore an audit error) if it can't?

Doesn't affect rendering in any way, so really just a semantics decision, but it does seem a bit weird to use `null` for what we're calling an unusual case.

(cue @patrickhulce with "of course it's not a gatherer error..." :)

Reply:
hahaha, of course it's not a gatherer error!

jk, not of course, but I do think it's not an error. It's a known limitation and I'd prefer to see us continue shrinking the size until it works and *then* throw an error if we fail, but I also see the rationale for throwing immediately instead. I don't feel particularly strongly though 🤷